### ISSUE TITLE
Getting an OpenClipboard Failed error

### ISSUE THREAD

**Issue by tjdennis** [Kir-Antipov/LineSorter#6]
I just installed your extension into Visual Studio 2019 (16.1.3) and doing a sort just returned an error.

OpenClipboard Failed (Exception from HRESULT: 0x800401D0 (CLIPBRD_E_CANT_OPEN))

![image](https://user-images.githubusercontent.com/4396052/59864881-fa6cbc00-9344-11e9-8acc-9bc4e606fa3f.png)

In case it's a conflict with another extension, this is what I have installed in my Visual Studio.

Microsoft Visual Studio Community 2019
Version 16.1.3
VisualStudio.16.Release/16.1.3+29009.5
Microsoft .NET Framework
Version 4.7.03190

Installed Version: Community

ASP.NET and Web Tools 2019   16.1.429.50124
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019   16.1.429.50124
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.1.429.50124
Azure App Service Tools v3.0.0

C# Tools   3.1.1-beta4-19281-06+58a4b1e79aea28115e66b06f850c83a3f1fcb6d3
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.1.77 (master@24013d5)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Favorite Documents   2.4.0
Favorites menu for Visual Studio IDE.

File Icons   2.7
Adds icons for files that are not recognized by Solution Explorer

File Nesting   2.6.85
Automatically nest files based on file name and enables developers to nest and unnest any file manually

GhostDoc   19.1.19150.0
Generate XML Comments from your code, maintain clean and up-to-date documentation, produce help documentation in multiple formats, use intelligent source code Spell Checker in Visual Studio.

HgSccPackage   2.0.6
Mercurial source code control package

JetBrains ReSharper Ultimate 2019.1.1   Build 191.0.20190501.122851
JetBrains ReSharper Ultimate package for Microsoft Visual Studio. For more information about ReSharper Ultimate, visit http://www.jetbrains.com/resharper. Copyright © 2019 JetBrains, Inc.

LineSorter Extension   1.0
LineSorter Visual Studio Extension Detailed Info

Markdown Editor   1.12.236
A full featured Markdown editor with live preview and syntax highlighting. Supports GitHub flavored Markdown.

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Mono Debugging for Visual Studio   16.1.1 (2473f22)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.1.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.61904.23160
Microsoft SQL Server Data Tools

Tabs Studio   4.5.0
The document tabs manager.

TypeScript Tools   16.0.10506.2004
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.1.1-beta4-19281-06+58a4b1e79aea28115e66b06f850c83a3f1fcb6d3
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.4 for F# 4.6   16.1.0-beta.19253.3+42526fe359672a05fd562dc16a91a43d0fe047a7
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.1.0.543 (d16-1@34a619991)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.1.0.418 (remotes/origin/d16-1@5b958bb10)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.2.112 (4db4af4)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.3.0.22 (HEAD/8e7764fdf)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2018-08@3cb36842fc4
    Java.Interop: xamarin/java.interop/d16-1@5ddc3e3
    LibZipSharp: grendello/LibZipSharp/d16-1@44de300
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.27.1@8212a2d
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-1@acabd26

Xamarin.iOS and Xamarin.Mac SDK   12.10.0.153 (750a879)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by Kira-NT** [comment]
Thanks for pointing out this error!

This behavior is caused by a simple line:
```csharp
Clipboard.SetText(text);
```
Apparently, you have installed some software that always keeps the clipboard open, so trying to install custom text there causes an exception...

At the moment I changed the call to clipboard API and added exception handling. Hope, it will fix this issue

**Comment by Kira-NT** [comment]
Just published an update ([LineSorter 4.7](https://marketplace.visualstudio.com/items?itemName=KirAntipov.LineSorterByKir))

Is your problem solved?

**Comment by tjdennis** [comment]
I found the other program causing the error. It was TightVNC which I use to keep an eye on one of my other development machines.  Exiting the viewer fixed the problem and the sort worked.

Your new update though might have caused some other problems that might affect everyone... Visual Studio hung for about 10 seconds before the sort, and then hung up for about 5 minutes after sorting.  It also scrambled the formatting of the text. I restarted Visual Studio and tried it a second time with the same results.

Original Text:
![image](https://user-images.githubusercontent.com/4396052/59880560-8d6c1d00-936a-11e9-9d6d-c0019e1459be.png)

Afterward, it looked like this 
![image](https://user-images.githubusercontent.com/4396052/59880615-a96fbe80-936a-11e9-8c1b-ac68b8f33a69.png)

This could be a result of Visual Studio trying to interpret and format the clipboard data as it's inserted.  Resharper might be in there as well.  

My test of your 4.6 version after stopping TightVnc didn't have the reformatting issue and worked instantly.  You may want to revert this change until you have a more complete solution and I'll just stop my TightVnc viewer when I need to sort something.  Everyone else may be getting this update that might cause them troubles too.

Thanks!

**Comment by tjdennis** [comment]
Here's my source file if it helps for testing.  It will probably show hundreds of bad references, but hopefully the sort can still work.


[WtUtilApp.zip](https://github.com/Kir-Antipov/LineSorter/files/3312130/WtUtilApp.zip)

**Comment by tjdennis** [comment]
Actually, I think TightVnc was still causing this problem. I was trying your update with it running when this happened.  Shutting it down fixed it again.

**Comment by Kira-NT** [comment]
In the current version I added exception handling when the clipboard is locked, so instead of working directly with the buffer, there is a [call to `EnvDTE.TextSelection.Text`](https://github.com/Kir-Antipov/LineSorter/commit/4f26115dd259dcbaedfc8a7cf8364ac6440d65cd#diff-41c6c7f1acbf8f7fcac92cac82a4e69aR95):

```csharp
((EnvDTE.TextSelection)dte.ActiveDocument.Selection).Text = text;
```

As long as the buffer is available, everything will work the same way as it was before, so don't worry about present LineSorter users!)
As you can see from the notes to the commit's code, I know that `TextSelection.Text` is very slow and buggy (I refused this API usage even before the first release). However, so far I don't know what else to do in case of a blocked clipboard...

**Comment by Kira-NT** [comment]
Tomorrow I'll try to reproduce your problem and look for solutions
It will simply be wrong to make you turn off your software forever, just to sort out a few lines, so don't worry, I'm sure I can come up with something!)

**Comment by tjdennis** [comment]
I've been experimenting some more.  I found that I can turn the clipboard transfer on or off in TightVNC.  Turning it off (which it never worked well enough to use anyway) fixes it for me.  

Using the TextSelection apparently works differently than the SetText or SetDataObject. The later just do a paste, where as TextSelection breaks it up into a hundred or more separate text changes - each causing VS and Resharper to apply formatting.  I made the mistake of trying to undo them all and it froze me again for 20 minutes so I killed the devenv process.

I think your current fix is good enough though. It should work for anyone with a locked clipboard and I can fix my problem by turning off the clipboard support.

Thanks for all your help!  I used to have to load the Menees Tools extension to get sorting but that was a lot of overkill that I never needed. And it doesn't provide all the options that yours does :)

**Comment by Kira-NT** [comment]
`TextSelection.Text` really works very-very strange! For the sake of interest, I measured its performance for different lengths of text and I can say that the insertion time is growing exponentially with each new letter... Apparently the text is inserted almost letter by letter
And it's kinda weird :D

I am very glad that your problem was resolved without having to turn off TightVNC. However, the very essence of the problem still bothers me, so I'll try to look for an adequate solution in any case!

Thank you for your feedback and for using LineSorter! :D
